### ISSUE TITLE
[Question] Exposing Raw Manifest File.

### ISSUE THREAD

**Issue by SemihGk** [google/ExoPlayer#4862]
I am quite new to ExoPlayer and so, I am sorry if the question is inappropriate. 

Basically, I need to get a raw manifest file from Exoplayer. However, I could not find any way to reach it. The only place that I can access to the manifest at [onTimelineChanged](https://google.github.io/ExoPlayer/doc/reference/com/google/android/exoplayer2/Player.EventListener.html#onTimelineChanged-com.google.android.exoplayer2.Timeline-java.lang.Object-int-) event. However, it is parsed manifest and there is no attribute which provides the raw manifest. The reason that I need raw manifest is for reporting purposes. We want to report the latest manifest when there is an error or rebuffering occurs. So, we cannot report the parsed attributes. That's why we need the raw manifest file.

My question is that is there any method that exposes this raw manifest file? If it is not, would you be interested in supporting this feature? Otherwise, I will have to expose this on my fork unless you can give me another suggestion. I will appreciate any advice.

Thank you,
Semih.

**Comment by erdemguven** [comment]
We don't keep the raw manifest file after it's parsed. (For DASH and SmoothStreaming) One way to keep manifest is providing it yourself to the MediaSource:

1. Download and store somewhere the manifest file.
1. Parse it using one of *ManifestParser classes.
1. Use DashMediaSource.Factory.createMediaSource(DashManifest) (or SmoothStreaming equivalent) to create your MediaSource.

**Comment by SemihGk** [comment]
Thank you for your response, @erdemguven . If it was only for VOD content, your case would be adequate but we would have to implement continuous download support for LIVE contents. Also, there is a risk that it may break things for some unknown reason in case of implementing a custom manifest downloader in our app. So, it may not worth to go for it.

I have another thought. I noticed that the HTTP library has callbacks on open and close methods such as onTransferStart and [onTransferEnd](https://github.com/google/ExoPlayer/blob/8cbcfd669db0595ad8c84425b7a99f62d55e3569/library/core/src/main/java/com/google/android/exoplayer2/upstream/DefaultHttpDataSource.java#L290) . I believe I can read the response to be able to get the raw manifest.

**Comment by erdemguven** [comment]
Also one alternative is passing a custom DataSource.Factory wrapper to DashMediaSource.Factory for manifestDataSourceFactory parameter so you can listen all calls for reading manifest. Keep in ming the DataSource created by manifestDataSourceFactory may be also used for reading UtcTimingElement of manifest from server.

**Comment by SemihGk** [comment]
Thank you. Yes, you are right. That's how I started it so far. I passed a custom DataSource factory through with a custom TransferListener class. I am able to listen to the events. However, all variables and methods which provides a readable stream/connection are either private or protected. It seems I have to use a hacky-ish solution to get them because I cannot see any other option.

**Comment by erdemguven** [comment]
Once you have custom DataSource (and a custom DataSource.Factory to create it) you can interfere all of the calls to the underlying real DataSource.
Also if in your case, there is no need for an extra access to server for UtcTimingElement then you can just use some existing classes. Something like this may work:
```
// create MediaSource
new DashMediaSource.Factory(new DefaultDashChunkSource.Factory(dataSourceFactory),
  new DataSource.Factory() {
    @Override
    public DataSource createDataSource() {
      lastDataSink = new ByteArrayDataSink();
      return new TeeDataSource(dataSourceFactory.createDataSource(), lastDataSink);
    }
  }).createMediaSource(uri);
// to get the content of the last read file
lastDataSink.getData();
```

**Comment by SemihGk** [comment]
I prefer not to override methods(especially the network calls) as much as I can just in case of a possible incompatibility issue when there is a new Exoplayer upgrade. Since this is a very simple feature, I do not want to make this too complicated. On the other hand, I finally got it working. I appreciate your support.

As for UTCTimingElement, it exists in our manifests and so there is no need to call it to the server. Plus, this would be very minor detail in addition to the manifest.

I did not notice ByteArrayDataSink class and it seems it is quite easier than what I have achieved so far. I will definitely try it. Thank you very much.
